### ISSUE TITLE
Request: Spelling check

### ISSUE THREAD

**Issue by strickvl** [charliermarsh/ruff#1628]
There are seemingly two big Python code checkers: [`pyspelling`](https://facelessuser.github.io/pyspelling/) and [`codespell`](https://github.com/codespell-project/codespell). Codespell doesn't catch everything, and pyspelling is slow to run. It'd be great to have a version / implementation of the logic of these (eventually including PySpelling's exception syntax etc) in `ruff`.

**Comment by colin99d** [comment]
@charliermarsh could we just port over typos, or is it too rust specific?

**Comment by charliermarsh** [comment]
(That is: call the typos functions directly in Rust, rather than using the CLI, and report typos as Ruff errors.)

**Comment by colin99d** [comment]
I have been thinking about this for a while and I think we should keep spell checking separate from Ruff for the following reasons:

1. This spell checker is already blazingly fast, so users of typos have no reason to switch
2. Spell checkers do not need to parse python code, so we would not save any additional time from them being part of the library
3. Having to support multiple languages could be burdensome

**Comment by charliermarsh** [comment]
Yeah that makes sense. Let's pass on this for now.

**Comment by MicaelJarniac** [comment]
> I have been thinking about this for a while and I think we should keep spell checking separate from Ruff for the following reasons:
> 
>     1. This spell checker is already blazingly fast, so users of typos have no reason to switch
> 
>     2. Spell checkers do not need to parse python code, so we would not save any additional time from them being part of the library
> 
>     3. Having to support multiple languages could be burdensome

Regarding 2, I think spell checkers can benefit greatly from parsing Python code.

For example, say I've installed a lib from PyPI that has typos on its function names. There's nothing I can do about them, they're outside my control, but I'll need to type the typos if I want to use these functions.

A normal spell checker, that treats everything as plain text, would see the typo and complain about it. But again, there's nothing I can do about them.

If the spell checker understands Python, it'll be able to tell what's my fault and what's outside my control, and so if the typo is on something from an external lib, it can ignore them, as I can't control that, but if the typo is on my own functions and variable names, it can then warn me.

**Comment by adam-grant-hendry** [comment]
> > I have been thinking about this for a while and I think we should keep spell checking separate from Ruff for the following reasons:
> > ```
> > 1. This spell checker is already blazingly fast, so users of typos have no reason to switch
> > 
> > 2. Spell checkers do not need to parse python code, so we would not save any additional time from them being part of the library
> > 
> > 3. Having to support multiple languages could be burdensome
> > ```
> 
> Regarding 2, I think spell checkers can benefit greatly from parsing Python code.
> 
> For example, say I've installed a lib from PyPI that has typos on its function names. There's nothing I can do about them, they're outside my control, but I'll need to type the typos if I want to use these functions.
> 
> A normal spell checker, that treats everything as plain text, would see the typo and complain about it. But again, there's nothing I can do about them.
> 
> If the spell checker understands Python, it'll be able to tell what's my fault and what's outside my control, and so if the typo is on something from an external lib, it can ignore them, as I can't control that, but if the typo is on my own functions and variable names, it can then warn me.

Also regarding 2, I need to spellcheck docstrings for my documentation, so I do actually do need to run in through python files sometimes.

**Comment by ashrub-holvi** [comment]
Good to keep in mind that, if I understood it correctly, there are two approaches for spellcheck - based on dictionary of correct words and based on dictionary/rules for mistakes. Both has own pros and cons. `Typos` and `codespell` is fixing known typos, `PySpelling` uses Aspell and Hunspell, I guess it's mostly dictionaries.
